lti_consumer/lti.py

```diff
@@ -0,0 +1,288 @@
+    https://www.imsglobal.org/specs/ltiv1p1
+    """
+    def __init__(self, xblock):
+        self.xblock = xblock
+
+    def get_signed_lti_parameters(self):
+        """
+        Signs LTI launch request and returns signature and OAuth parameters.
+
+        Arguments:
+            None
+
+        Returns:
+            dict: LTI launch parameters
+        """
+
+        # Must have parameters for correct signing from LTI:
+        lti_parameters = {
+            u'user_id': self.xblock.user_id,
+            u'oauth_callback': u'about:blank',
+            u'launch_presentation_return_url': '',
+            u'lti_message_type': u'basic-lti-launch-request',
+            u'lti_version': 'LTI-1p0',
+            u'roles': self.xblock.role,
+
+            # Parameters required for grading:
+            u'resource_link_id': self.xblock.resource_link_id,
+            u'lis_result_sourcedid': self.xblock.lis_result_sourcedid,
+
+            u'context_id': self.xblock.context_id,
+        }
+
+        if self.xblock.has_score:
+            lti_parameters.update({
+                u'lis_outcome_service_url': self.xblock.outcome_service_url
+            })
+
+        self.xblock.user_email = ""
+        self.xblock.user_username = ""
+
+        # Username and email can't be sent in studio mode, because the user object is not defined.
+        # To test functionality test in LMS
+
+        if callable(self.xblock.runtime.get_real_user):
+            real_user_object = self.xblock.runtime.get_real_user(self.xblock.runtime.anonymous_student_id)
+            try:
+                self.xblock.user_email = real_user_object.email
+            except AttributeError:
+                self.xblock.user_email = ""
+            try:
+                self.xblock.user_username = real_user_object.username
+            except AttributeError:
+                self.xblock.user_username = ""
+
+        if self.xblock.ask_to_send_username and self.xblock.user_username:
+            lti_parameters["lis_person_sourcedid"] = self.xblock.user_username
+        if self.xblock.ask_to_send_email and self.xblock.user_email:
+            lti_parameters["lis_person_contact_email_primary"] = self.xblock.user_email
+
+        # Appending custom parameter for signing.
+        lti_parameters.update(self.xblock.prefixed_custom_parameters)
+
+        headers = {
+            # This is needed for body encoding:
+            'Content-Type': 'application/x-www-form-urlencoded',
+        }
+
+        key, secret = self.xblock.lti_provider_key_secret
+        oauth_signature = get_oauth_request_signature(key, secret, self.xblock.launch_url, headers, lti_parameters)
+
+        # Parse headers to pass to template as part of context:
+        oauth_signature = dict([param.strip().replace('"', '').split('=') for param in oauth_signature.split(',')])
+
+        oauth_signature[u'oauth_nonce'] = oauth_signature[u'OAuth oauth_nonce']
+        del oauth_signature[u'OAuth oauth_nonce']
```

Comment:
Style nit: You could simply use

    oauth_signature[u'oauth_nonce'] = oauth_signature.pop(u'OAuth oauth_nonce')

To get the value and remove the key from the dictionary in one go.